### PR TITLE
[None][feat] Upgrade transformers dependency to 5.3.0

### PR DESCRIPTION
**Pull request by longlee0622**
# [None][feat] Upgrade transformers dependency to 5.3.0

## Summary

Bumps `transformers` from 4.x to **5.3.0** and adapts TensorRT-LLM to the wide set of API removals, renames, and behavioral changes in the v5 release. This is a large cross-cutting change (116 files, +1521 / -560). The branch is intentionally split into ~96 small, single-concern commits to make review tractable; the plan is to **squash on merge**, so individual commit messages do not need to be final.

Reviewers: this PR description is the canonical log of what changed and why. The per-commit history is a review aid — please use it to step through one concrete issue at a time rather than reading the full diff in one pass.

---

## What changed in transformers 5.3.0

The v5 release is a breaking release. The classes of changes we hit:

### 1. RoPE config refactor
- `config.rope_scaling` is now a **computed property** that delegates to a new `rope_parameters` dict.
- The dict uses `"rope_type"` instead of `"type"`.
- `"rope_type": "default"` now explicitly means "no scaling" — many models' custom `_init_rope` code doesn't recognise it.
- `rope_theta` can live inside `rope_parameters` only (not as a top-level attribute).
- Some models (Gemma3) moved to **per-layer-type** RoPE config.
- VL models keep RoPE fields on `text_config`, not the outer config.

### 2. Tokenizer API changes
- `batch_encode_plus` removed — `__call__` has the same signature.
- `vocab_size` property and `__len__` may raise `NotImplementedError` on some tokenizers (Qwen2Tokenizer).
- `apply_chat_template` default return type changed (needs `return_dict=False`).
- **LlamaTokenizer in v5 silently replaces `tokenizer.json`'s `ByteLevel` pre-tokenizer with `Metaspace`** when the repo declares `tokenizer_class: LlamaTokenizer` — this strips spaces from prompts and caused a catastrophic accuracy drop on DeepSeek-V3 (~25% vs ~63% on GSM8K) until fixed.

### 3. Fused MoE expert weights
- Mixtral, Qwen3Moe, Qwen3Next, Qwen3VL now ship experts as **fused tensors** rather than per-expert weights.
- `gate_up_proj` is a single tensor that must be split into `w1` / `w3` for TRT-LLM.
- FP8 scale tensors are also fused and must be unfused.
- Expert weight keys dropped the `.weight` suffix in some places.
- MoE gate `router_logits` is now a tuple in v5.
- `top_k` attribute renamed on some MoE modules.

### 4. Config structure moves
- `MllamaConfig.pad_token_id` moved to `text_config`.
- `tie_word_embeddings` no longer a default on all configs (e.g. `CLIPVisionConfig`).
- Qwen2.5-VL / Qwen3-VL: `max_position_embeddings`, `rms_norm_eps`, `head_dim` all live on `text_config` / inner configs.

### 5. Removed APIs needing replacement
| Removed | Replacement |
|---|---|
| `modeling_utils.no_init_weights` | `model.init_weights()` lazily |
| `modeling_utils.load_sharded_checkpoint` | local shard loader |
| `modeling_utils.get_parameter_device` / `get_parameter_dtype` | `ModuleUtilsMixin`-style inline helpers |
| `cache_utils.HybridCache` | `StaticCache` |
| `cache_utils.SlidingWindowCache` | merged into `StaticCache` (aliased for custom model code) |
| `DynamicCache.from_legacy_cache` / `.to_legacy_cache` | manual loop via `cache.update()` / layer iteration |
| `integrations.executorch.sdpa_mask_without_vmap` | `functools.partial(masking_utils.sdpa_mask, use_vmap=False)` |

### 6. Export / AutoDeploy-specific breakage
- `find_packed_sequence_indices` now calls `.all()` on attention mask → fails on meta tensors during export.
- `BambaModel._update_causal_mask` removed.
- `torch.is_autocast_enabled()` assumptions changed for unknown device types.

### 7. Ecosystem
- `mistral-common` needs to be bumped to `1.9.1` for v5 compatibility.

---

## How we fixed it

Fixes are grouped here by theme; the branch has one commit per concrete issue if you want the per-case explanation.

### Accuracy-critical fixes (highlighted)
- **ByteLevel tokenizer reload** (`tokenizer.py`): detect the Metaspace / `tokenizer.json` ByteLevel mismatch after `AutoTokenizer.from_pretrained` and reload via `PreTrainedTokenizerFast` to respect `tokenizer.json`. Authored by **@dc3671**. Unblocked DeepSeek-V3 and LlamaTokenizer-declaring ByteLevel repos.
- **ByteLevel fix in `kv_cache_aware` router**: disaggregated serving's router pre-tokenizes prompts and overwrites `request.prompt` with token IDs — it was bypassing the tokenizer workaround and reintroducing the space-stripping bug in disagg CI. Wired the workaround into `serve/router.py::BlockHashMixin._get_tokenizer`.
- **Preserve `rope_theta` when clearing default `rope_scaling`**: `rope_scaling` delegates to `rope_parameters`; clearing the latter without copying `rope_theta` dropped theta to the default 10000 and caused ~14% accuracy drops on Qwen3-8B / Qwen3-30B-A3B.

### RoPE refactor adaptation
- Config loader extracts `rope_scaling` / `rope_theta` from `rope_parameters`.
- `rope_type == "default"` is treated as "no scaling" (maps to `rope_gpt_neox` in `PositionEmbeddingType`).
- Phi-3's custom `_init_rope` clears `rope_scaling` entirely when `"default"` (can't interpret it).
- `type` / `rope_type` keys kept in sync for custom HF code paths.
- VL models apply the same RoPE handling to `text_config`.
- Per-layer-type RoPE (Gemma3) supported end-to-end.
- Pixtral: populate missing `rope_parameters` for `PixtralRotaryEmbedding`.
- DeepSeek V3 yarn RoPE override now unconditional.

### MoE weight unfusing
- Generic path: unfuse Mixtral / Qwen3Moe / Qwen3Next / Qwen3VL fused expert weights in weight mappers.
- Split `gate_up_proj` into `w1` / `w3`.
- Unfuse FP8 scale tensors.
- Remap keys that lost the `.weight` suffix.
- Guard against non-iterable fused expert modules in MoE patches.
- Handle tuple `router_logits` and renamed `top_k` attr.
- KimiK2 equivalence tests updated to match new fused key layout.

### Removed-API replacements
- `batch_encode_plus` → `__call__` in tokenizer wrapper and callers.
- `HybridCache` → `StaticCache` helper in tests.
- `SlidingWindowCache` aliased to `StaticCache` on load of Phi-4 multimodal's custom code.
- `DynamicCache` legacy-tuple helpers in TRT attention tests.
- `no_init_weights` removed — tests construct eagerly and call `init_weights()`.
- `load_sharded_checkpoint` reimplemented locally for Llama4 vision encoder.
- `get_parameter_device` / `get_parameter_dtype` — inline `ModuleUtilsMixin` equivalent.
- `sdpa_mask_without_vmap` → `partial(sdpa_mask, use_vmap=False)`.

### Multimodal / VL
- Qwen2.5-VL: resolve `max_position_embeddings`, `rms_norm_eps`, `head_dim`, RoPE fields from `text_config`.
- Qwen3-VL: transpose MoE expert weights, fix weight mapper fused-weight handling, video MRope position IDs, `_check_and_adjust_experts_implementation` override.
- LlavaNext dtype mismatch fix.
- VilaMultimodalProjector v5 compat.
- Mistral3 VLM test fixes.
- Removed `mm_token_type_ids` from HF inputs where v5 rejects it.

### AutoDeploy / export
- SDPA mask patch fallback when executorch helper is removed.
- Patch `BambaModel` when `_update_causal_mask` is absent.
- Patch `find_packed_sequence_indices` for meta tensors.
- Patch `torch.is_autocast_enabled` for unknown device types.

### Tokenizer / inputs
- `vocab_size` resolution guarded with `try/except NotImplementedError`, with fallbacks through `len(tokenizer)` and inner tokenizer.
- `apply_chat_template` callers pass `return_dict=False` where needed.

### Shim cleanup
After stabilising on v5, removed v4 fallback branches that became unreachable (since `requirements.txt` pins `>=5.3.0`): `hasattr(…, 'batch_encode_plus')`, `HybridCache` try/except, `SlidingWindowCache` hasattr, `DynamicCache` legacy-method hasattr, `no_init_weights` getattr, `get_parameter_device/dtype` try/except, a dead `transformers < 4.53.0` version check. v5-correct guards (e.g. `getattr` for `tie_word_embeddings` absence, `try/except NotImplementedError` for `vocab_size`) are kept — those aren't fallbacks, they encode real v5 behavior.

### CI / tests
- Unwaive disagg tests that were temporarily waived for the upgrade.
- Disable two-model Eagle3 test (feature deprecated).
- AutoDeploy test fixes for v5.

---

## Accuracy validation

| Model | Before | After | Notes |
|---|---|---|---|
| DeepSeek-V3-Lite (GSM8K) | ~25% | ~63% | ByteLevel tokenizer fix |
| Qwen3-8B | 73% | 87% | `rope_theta` preservation |
| Qwen3-30B-A3B | 57% | 83% | `rope_theta` preservation |

These regressions were silent (no exception, just garbage output) — discovered only via accuracy sweeps.

---

## Testing

- L0 unit tests: pass locally
- Pre-merge CI: pass (pending final run)

---

## Credits

Thanks to contributors who shipped fixes on this branch:

- **@dc3671** — LlamaTokenizer ByteLevel reload fix (the biggest accuracy recovery in this PR).
- **@lucaslie** — AutoDeploy transformers 5.3 fixes and `export_to_onnx` deprecation scheduling.
- **@brb-nv** — Unwaive disagg tests for the upgrade.
- **@eopXD** — Gemma 3 nested `rope_parameters` resolution.
- **@ziyixiong-nv** — Disable two-model Eagle3 in CI.
- **@2ez4bz** — Multimodal fixes.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/export/library/transformers_causal_mask.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+# SPDX-FileCopyrightText: Copyright (c) 2022-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (longlee0622) [inline tensorrt_llm/_torch/auto_deploy/export/library/transformers_causal_mask.py]:
```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

**Review by lucaslie — APPROVED** [review]
approving on behalf of AutoDeploy

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 2]
```diff
@@ -484,10 +484,16 @@ def from_config(config) -> "RopeParams":
...
-            assert not set(hf_rope_parameters.keys()).issubset(
-                ALLOWED_ATTENTION_LAYER_TYPES), (
-                    "Per-layer-type RoPE configuration is not supported yet.")
-            config.update(hf_rope_parameters)
+            if set(hf_rope_parameters.keys()).issubset(
+                    ALLOWED_ATTENTION_LAYER_TYPES):
+                # Per-layer-type RoPE config (e.g. Gemma3 in transformers 5.x).
+                # Pick "full_attention" as the default; callers override theta
+                # for sliding-window layers independently.
+                flat = hf_rope_parameters.get(
```
Comment (eopXD) [inline tensorrt_llm/_torch/attention_backend/interface.py]: Lets emit a warning log here explicitly for clearance. Silent fall back here may trouble us in debugging in the future.

File: tensorrt_llm/_torch/pyexecutor/config_utils.py [hunk 3]
```diff
@@ -283,4 +290,25 @@ def load_pretrained_config(model_name_or_path: str,
...
+    if isinstance(rope_scaling, dict):
+        rope_type = rope_scaling.get("rope_type", rope_scaling.get("type"))
+        has_real_scaling = any(k for k in rope_scaling
+                               if k not in ("rope_type", "type", "rope_theta"))
+        if rope_type == "default" and not has_real_scaling:
+            # Preserve rope_theta before clearing, since rope_parameters
+            # (which rope_scaling delegates to) will also become None and
+            # rope_theta may only exist there in transformers 5.x.
+            rope_theta = rope_scaling.get("rope_theta")
+            if rope_theta is not None:
```
Comment (eopXD) [inline tensorrt_llm/_torch/pyexecutor/config_utils.py:315]: Do we need else-case here for error handling and future upgrade?

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 4]
```diff
@@ -66,7 +66,14 @@ def __init__(
...
         self.attention_window_size = None
         if is_sliding:
-            rope_params.theta = config.rope_local_base_freq
+            # transformers 5.x moved rope_local_base_freq into
+            # rope_parameters["sliding_attention"]["rope_theta"]
+            local_freq = getattr(config, 'rope_local_base_freq', None)
+            if local_freq is None:
+                rp = getattr(config, 'rope_parameters', {})
+                local_freq = rp.get('sliding_attention',
+                                    {}).get('rope_theta', 10000.0)
```
Comment (eopXD) [inline tensorrt_llm/_torch/models/modeling_gemma3.py]: Should we raise instead of providing a fallback? This concern should apply to other similar logic on other `modeling_*` scripts.

**Review by eopXD — COMMENTED** [review]
Some nits on details.

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 2 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/attention_backend/interface.py]: thanks for the feedback. added one more commit to address your comments. please help to review.

Comment (eopXD) [reply tensorrt_llm/_torch/attention_backend/interface.py]: Looks good. Thank you for the update.

**Review by eopXD — APPROVED** [review]
Looks good to me.

**Review by nv-guomingz — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 5]
```diff
@@ -1075,4 +1075,4 @@ def forward(
         return hidden_states
 
 
-AutoConfig.register(NemotronHConfig.model_type, NemotronHConfig)
+AutoConfig.register(NemotronHConfig.model_type, NemotronHConfig, exist_ok=True)
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py]: I think you can remove NemotronHConfig class above, and import NemotronHConfig from transformers

File: tensorrt_llm/_torch/models/modeling_exaone4.py [hunk 6]
```diff
@@ -29,7 +29,7 @@
     class Exaone4Config(PretrainedConfig):
         model_type = "exaone4"
 
-    AutoConfig.register(Exaone4Config.model_type, Exaone4Config)
+    AutoConfig.register(Exaone4Config.model_type, Exaone4Config, exist_ok=True)
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_exaone4.py]: Same here

File: tensorrt_llm/_torch/models/modeling_exaone_moe.py [hunk 7]
```diff
@@ -53,7 +53,7 @@ class ExaoneMoEConfig(PretrainedConfig):
     "Register ExaoneMoEConfig to mimic the ExaoneMoE model.",
     key="EXAONE_MOE_REGISTER_WARNING"
 )
-AutoConfig.register(ExaoneMoEConfig.model_type, ExaoneMoEConfig)
+AutoConfig.register(ExaoneMoEConfig.model_type, ExaoneMoEConfig, exist_ok=True)
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_exaone_moe.py]: Same here

**Review by yechank-nvidia — APPROVED** [review]
nit) You can remove the `register` and instead import model's config from transformers.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py (hunk 5 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/models/modeling_nemotron_h.py]: good catch. We are already on 5.x path, the import should just work. updated in my latest commit

**Review by zhenhuaw-me — APPROVED** [review]

**Review by QiJune — APPROVED** [review]

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "GB200-12_GPUs-3_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE2-GPU8-Post-Merge-7,GB200-20_GPUs-5_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE4-GPU16-Post-Merge-2"

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "GB200-16_GPUs-4_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE2-GPU8-GEN1-NODE2-GPU8-Post-Merge-1,GB200-16_GPUs-4_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE2-GPU8-GEN1-NODE2-GPU8-Post-Merge-2,GB200-12_GPUs-3_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE2-GPU8-Post-Merge-7,GB200-20_GPUs-5_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE4-GPU16-Post-Merge-2"

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "GB200-16_GPUs-4_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE2-GPU8-GEN1-NODE2-GPU8-Post-Merge-1,GB200-16_GPUs-4_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE2-GPU8-GEN1-NODE2-GPU8-Post-Merge-2,GB200-12_GPUs-3_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE2-GPU8-Post-Merge-7,GB200-20_GPUs-5_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE4-GPU16-Post-Merge-2"

**Review by tburt-nv — APPROVED** [review]
The only transitive dependency addition appears to be typer, which is MIT licensed.

**Review by Tabrizian — APPROVED** [review]
